File: problems/1249_minimum_remove_to_make_valid_parentheses.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        depth = 0
        i = 0
        while i < len(s):
            if s[i] == "(":
                depth += 1
                i += 1
            elif s[i] == ")" and depth > 0:
                depth -= 1
                i += 1
            elif s[i] == ")" and depth == 0:
                s = s[0:i] + s[i + 1 :]
            else:
                i += 1

        depth = 0
        i = len(s) - 1
        while i >= 0:
            if s[i] == ")":
                depth += 1
            elif s[i] == "(" and depth > 0:
                depth -= 1
            elif s[i] == "(" and depth == 0:
                s = s[0:i] + s[i + 1 :]

            i -= 1

        return s

File: problems/1249-minimum-remove-to-make-valid-parentheses.ts
function minRemoveToMakeValid(s: string): string {
  let depth = 0
  let i = 0

  while (i < s.length) {
    if (s[i] === "(") {
      depth++
      i++
    } else if (s[i] === ")" && depth > 0) {
      depth--
      i++
    } else if (s[i] === ")" && depth === 0) {
      s = s.slice(0, i) + s.slice(i + 1)
    } else {
      i++
    }
  }

  depth = 0
  i = s.length - 1

  while (i >= 0) {
    if (s[i] === ")") {
      depth++
    } else if (s[i] === "(" && depth > 0) {
      depth--
    } else if (s[i] === "(" && depth === 0) {
      s = s.slice(0, i) + s.slice(i + 1)
    }
    i--
  }

  return s
}
